packages/integration-sdk-runtime/src/synchronization/batchBySize.ts

```diff
@@ -0,0 +1,89 @@
+import {
+  Entity,
+  IntegrationLogger,
+  Relationship,
+} from '@jupiterone/integration-sdk-core';
+import { UploadDataLookup } from '.';
+import { MAX_BATCH_SIZE } from './shrinkBatchRawData';
+const BATCH_THRESHOLD = 0.9; //will stop chunking if chunk size is between BATCH_THRESHOLD*sizeInBytes and sizeInBytes
+
+export function chunkBySize<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  sizeInBytes: number,
+  logger: IntegrationLogger,
+): T[K][][] {
+  return chunk(data, sizeInBytes, logger);
+}
+export function chunk<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  sizeInBytes: number,
+  logger: IntegrationLogger,
+): T[K][][] {
+  const chunkedData: T[K][][] = [];
+  if (getSizeOfObject(data) < sizeInBytes) {
+    return [data]; //Just one chunk of all data
+  }
+  let bestIndex = linearSearch(data, sizeInBytes);
+  if (bestIndex <= 0) {
+    //If the first entity is too big
+    handleSearchError(data, sizeInBytes, logger); // we remove rawdata
+    bestIndex = 1; //and send it to the persister alone. TODO: find a way to avoid sending this entities as a single call. It shouldn't be an issue since this only happens if there is a single entity above sizeInBytes
+  }
+  chunkedData.push(data.slice(0, bestIndex));
+  if (bestIndex !== data.length)
+    chunkedData.push(
+      ...chunk(data.slice(bestIndex), sizeInBytes, logger), //chunk the rest of the data
+    );
+  return chunkedData;
+}
+//We are linearly searching for the best index
+function linearSearch<T extends UploadDataLookup, K extends keyof T>(
```

Comment:
Formatting nit: make sure to leave a space between functions, and always prefer docstrings over comments when describing the intent of a function.

```suggestion

/** 
 * We are linearly searching for the best index
 */
function linearSearch<T extends UploadDataLookup, K extends keyof T>(
```